helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -632,31 +636,33 @@ private void subscribeForChanges(NotificationContext.Type callbackType, String p
             case CUSTOMIZED_VIEW:
             case TARGET_EXTERNAL_VIEW: {
               // check if bucketized
-              BaseDataAccessor<ZNRecord> baseAccessor = new ZkBaseDataAccessor<>(_zkClient);
-              List<ZNRecord> records = baseAccessor.getChildren(path, null, 0, 0, 0);
-              for (ZNRecord record : records) {
-                HelixProperty property = new HelixProperty(record);
-                String childPath = path + "/" + record.getId();
-
-                int bucketSize = property.getBucketSize();
-                if (bucketSize > 0) {
-                  // subscribe both data-change and child-change on bucketized parent node
-                  // data-change gives a delete-callback which is used to remove watch
-                  List<String> bucketizedChildNames = subscribeChildChange(childPath, callbackType);
+              if (BUCKETIZE_ZNRECORD_ENABLED) {
+                BaseDataAccessor<ZNRecord> baseAccessor = new ZkBaseDataAccessor<>(_zkClient);
+                List<ZNRecord> records = baseAccessor.getChildren(path, null, 0, 0, 0);
+                for (ZNRecord record : records) {
+                  HelixProperty property = new HelixProperty(record);
+                  String childPath = path + "/" + record.getId();
+                  int bucketSize = property.getBucketSize();
+                  // Do data-change subscribe for both bucketized parent node and non-bucketized node.
+                  // For bucketized parent node, data-change gives a delete-callback to remove watch.
                   subscribeDataChange(childPath, callbackType);
-
-                  // subscribe data-change on bucketized child
-                  if (bucketizedChildNames != null) {
-                    for (String bucketizedChildName : bucketizedChildNames) {
-                      String bucketizedChildPath = childPath + "/" + bucketizedChildName;
-                      subscribeDataChange(bucketizedChildPath, callbackType);
+                  if (bucketSize > 0) {
+                    // subscribe child-change on bucketized parent node.
+                    List<String> bucketizedChildNames =
+                        subscribeChildChange(childPath, callbackType);
+
+                    // subscribe data-change on bucketized child
+                    if (bucketizedChildNames != null) {
+                      for (String bucketizedChildName : bucketizedChildNames) {
+                        String bucketizedChildPath = childPath + "/" + bucketizedChildName;
+                        subscribeDataChange(bucketizedChildPath, callbackType);
+                      }
                     }
                   }
-                } else {
-                  subscribeDataChange(childPath, callbackType);
                 }
+                break;
               }
-              break;
+              // go to default branch if bucketized feature is not enabled.
```

Comment:
This is kind of tricky: readability is not very straightforward and it is easy to make a mistake if later someone adds another case after this, it'll break the logic.

How about this: we wrap the check into a private method and make it not easy to make a mistake?
```
shouldReadBucketizedChildren(type) {
  Set<> types; // CURRENT_STATE, ..., TARGET_EXTERNAL_VIEW
  return types.contains(type) && BUCKETIZE_ZNRECORD_ENABLED;
}

// Then for the block
if (shouldReadBucketizedChildren(_changeType) {
  // bucketized subscribe block
} else {
  // default block
}
```

Reply:
Updated.